sources/gpio_driver.c

```diff
@@ -0,0 +1,190 @@
+/**
+ * @file gpio_driver.c
+ * 
+ * @author Sam Donnelly (samueldonnelly11@gmail.com)
+ * 
+ * @brief GPIO functions
+ * 
+ * @version 0.1
+ * @date 2022-05-27
+ * 
+ * @copyright Copyright (c) 2022
+ * 
+ */
+
+
+//=======================================================================================
+// Includes 
+
+#include "gpio_driver.h"
+
+//=======================================================================================
+
+
+//=======================================================================================
+// Initialization 
+
+// GPIO communication initialization 
+void gpio_port_init(void)
+{
+    // TODO create a clocks driver to enable these and other clocks (RCC) 
+    
+    // Enable GPIOA clock - RCC_AHB1ENR register, bit 0 
+    RCC->AHB1ENR |= (SET_BIT << SHIFT_0);
+
+    // Enable GPIOB clock - RCC_AHB1ENR register, bit 1 
+    RCC->AHB1ENR |= (SET_BIT << SHIFT_1);
+
+    // Enable GPIOC clock - RCC_AHB1ENR register, bit 2 
+    RCC->AHB1ENR |= (SET_BIT << SHIFT_2);
+
+    // Enable GPIOD clock - RCC_AHB1ENR register, bit 3 
+    RCC->AHB1ENR |= (SET_BIT << SHIFT_3);
+}
+
+
+// GPIO pin initialization 
+void gpio_pin_init(
+    GPIO_TypeDef  *gpio, 
+    pin_selector_t pin_num,
+    gpio_moder_t   moder,
+    gpio_otyper_t  otyper,
+    gpio_ospeedr_t ospeedr,
+    gpio_pupdr_t   pupdr)
+{
+    // Set the mode 
+    gpio_moder(gpio, moder, pin_num); 
+
+    // Set the output type 
+    gpio_otyper(gpio, otyper, pin_num); 
+
+    // Set the output speed 
+    gpio_ospeedr(gpio, ospeedr, pin_num); 
+
+    // Set as pull-up or pull-down 
+    gpio_pupdr(gpio, pupdr, pin_num); 
+}
+
+//=======================================================================================
+
+
+//=======================================================================================
+// Write functions 
+
+// TODO see if the ODR register can be used instead of the BSRR register 
```

Comment:
I always recommend creating a ticket/feature card and linking it in the code with any TODOs.

Not really relevant here, but something great you can do in the future